extensions/amp-consent/0.1/amp-consent.js

```diff
@@ -18,34 +18,199 @@ import {CSS} from '../../../build/amp-consent-0.1.css';
+    this.consentUIRequired_ = {};
   }
 
   buildCallback() {
-    if (!isExperimentOn(this.ampdoc.win, AMP_CONSENT_EXPERIMENT)) {
+    if (!isExperimentOn(this.win, AMP_CONSENT_EXPERIMENT)) {
       return;
     }
+
+    // TODO: Decide what to do with incorrect configuration.
+    this.assertAndParseConfig_();
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_POLICY_MANGER)
+        .then(manager => {
+          this.consentPolicyManager_ = manager;
+          this.generateDefaultPolicy_();
+          const policyKeys = Object.keys(this.policyConfig_);
+          for (let i = 0; i < policyKeys.length; i++) {
+            this.consentPolicyManager_.registerConsentPolicyInstance(
+                policyKeys[i], this.policyConfig_[policyKeys[i]]);
+          }
+        });
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_STATE_MANAGER)
+        .then(manager => {
+          this.consentStateManager_ = manager;
+          this.init_();
+        });
+
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout == Layout.NODISPLAY;
   }
+
+  /**
+   * Init the amp-consent by registering and initiate consent instance.
+   */
+  init_() {
+    const instanceKeys = Object.keys(this.consentConfig_);
+    for (let i = 0; i < instanceKeys.length; i++) {
+      const instanceId = instanceKeys[i];
+      this.consentStateManager_.registerConsentInstance(instanceId);
+      this.getConsentRemote_(instanceId).then(response => {
+        this.parseConsentResponse_(instanceId, response);
+        this.handleUI_(instanceId);
+      }).catch(unusedError => {
+        // TODO: Handle errors
+      });
+    }
+  }
+
+  /**
+   * Generate default consent policy if not defined
+   */
+  generateDefaultPolicy_() {
+    if (this.policyConfig_ && this.policyConfig_['default']) {
+      return;
+    }
+    // Generate default policy
+    const instanceKeys = Object.keys(this.consentConfig_);
+    const defaultWaitForItems = {};
+    for (let i = 0; i < instanceKeys.length; i++) {
+      // TODO: Need to support an array.
+      defaultWaitForItems[instanceKeys[i]] = undefined;
+    }
+    const defaultPolicy = {
+      'waitFor': defaultWaitForItems,
+    };
+    this.policyConfig_['default'] = defaultPolicy;
+  }
+
+  /**
+   * Get localStored consent info, and send request to get consent from endpoint
+   * @param {string} instanceId
+   * @return {!Promise<!JsonObject>}
+   */
+  getConsentRemote_(instanceId) {
+    // Note: Expect the request to look different in following versions.
+    const request = /** @type {!JsonObject} */ ({
+      'consentInstanceId': instanceId,
+    });
+    const init = {
+      credentials: 'include',
+      method: 'POST',
+      body: request,
+      requireAmpResponseSourceOrigin: false,
+    };
+    const href =
+        this.consentConfig_[instanceId]['checkConsentHref'];
+    assertHttpsUrl(href, this.element);
+    return Services.xhrFor(this.win)
+        .fetchJson(href, init)
```

Comment:
Privacy leak! This is called as a result of `#buildCallback`, and there don't appear to be any restrictions on `href`, allowing me to POST to my server!

Reply:
!! I'll fix this by waiting for document to be first visible.

Reply:
https://github.com/ampproject/amphtml/pull/14656